c/tskit/tables.c

```diff
@@ -7799,6 +7799,182 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+                ret = tsk_population_table_add_row(
+                    &self->populations, pop.metadata, pop.metadata_length);
+                if (ret < 0) {
+                    goto out;
+                }
+                population_map[pid] = ret;
+            }
+            pid = population_map[pid];
+        }
+        ret = tsk_node_table_add_row(&self->nodes, node.flags, node.time, pid, iid,
+            node.metadata, node.metadata_length);
+        if (ret < 0) {
+            goto out;
+        }
+        if (ret != k) {
+            goto out;
+        }
+        node_map[nid] = ret;
+    }
+
+    // edges
+    for (eid = 0; eid < (tsk_id_t) tables.edges.num_rows; eid++) {
+        tsk_edge_table_get_row(&tables.edges, eid, &edge);
+        pid = node_map[tables.edges.parent[eid]];
+        cid = node_map[tables.edges.child[eid]];
+        if ((pid != TSK_NULL) && (cid != TSK_NULL)) {
+            ret = tsk_edge_table_add_row(&self->edges, edge.left, edge.right, pid, cid,
+                edge.metadata, edge.metadata_length);
+            if (ret < 0) {
+                goto out;
+            }
+        }
+    }
+
+    // mutations and sites
+    mid = 0;
+    for (sid = 0; sid < (tsk_id_t) tables.sites.num_rows; sid++) {
+        tsk_site_table_get_row(&tables.sites, sid, &site);
+        while ((mid < (tsk_id_t) tables.mutations.num_rows)
+               && (tables.mutations.site[mid] == sid)) {
+            tsk_mutation_table_get_row(&tables.mutations, mid, &mut);
+            nid = node_map[mut.node];
+            if (nid != TSK_NULL) {
+                if (site_map[sid] == TSK_NULL) {
+                    ret = tsk_site_table_add_row(&self->sites, site.position,
+                        site.ancestral_state, site.ancestral_state_length, site.metadata,
+                        site.metadata_length);
+                    if (ret < 0) {
+                        goto out;
+                    }
+                    site_map[sid] = ret;
+                }
+                pid = mut.parent;
+                if (pid != TSK_NULL) {
+                    pid = mutation_map[pid];
+                }
+                ret = tsk_mutation_table_add_row(&self->mutations, site_map[sid], nid,
+                    pid, mut.derived_state, mut.derived_state_length, mut.metadata,
+                    mut.metadata_length);
+                if (ret < 0) {
+                    goto out;
+                }
+                mutation_map[mid] = ret;
+            }
+            mid++;
+        }
+    }
+
+    // migrations
+    for (gid = 0; gid < (tsk_id_t) tables.migrations.num_rows; gid++) {
+        tsk_migration_table_get_row(&tables.migrations, gid, &mig);
+        nid = node_map[mig.node];
+        if (nid != TSK_NULL) {
+            sid = mig.source;
+            did = mig.dest;
+            if (sid != TSK_NULL) {
+                sid = population_map[sid];
+            }
+            if (did != TSK_NULL) {
```

Comment:
Actually, what if this is a population we haven't already seen before? You would need to check the map, I guess, and update if necessary. I don't think I'd bother though: unless you have an application using migrations, I would just raise an error if there's stuff in the migration table. I don't think it's really used much. This is what we do in simplify fwiw (there's an error code for "we don't support migrations" which could be reused).

Reply:
Right, the dest population has to be in the map, but it is possible the source isn't. I didn't know source and dest had to be non-NULL. My idea was that in this case the source pop would be NULL in the subset tables. Alternatively we could remove migrations with source pop not seen in any of the `nodes` from the subset. I think we are so close to being able to support the Migrations Table in this operation that we should just go for it.

Reply:
I think either of them could be missing from the map. Suppose you have a node that starts in one population then goes for a long journey through hundreds of populations. In that case the majority of the populations will not have been seen before. So, I think we'd have to append all these populations to the map.

There's a fair bit of testing to be done in this... It would be a lot quicker the other way.

Reply:
Ah, of course - I'd intended them to be NULL in this case, but of course it doesn't make sense for them to be NULL. I agree that throwing an error is the thing to do.

Reply:
Whoops - i meant to throw an error if the source or dest turns out to be unmapped.

Reply:
I decided to remove support for migrations altogether...